runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/StreamingEngineClient.java

```diff
@@ -0,0 +1,452 @@
+        newConnectionsState,
+        connections.get());
+
+    isBudgetRefreshPaused.set(false);
+    connections.set(newConnectionsState);
+
+    // On first worker metadata. Trigger
+    if (getWorkerMetadataReady.getCount() > 0) {
+      getWorkerMetadataReady.countDown();
+    } else {
+      requestBudgetRefresh();
+    }
+  }
+
+  public ImmutableList<Long> getAndResetThrottleTimes() {
+    StreamEngineConnectionState currentConnections = connections.get();
+
+    ImmutableList<Long> keyedWorkStreamThrottleTimes =
+        currentConnections.windmillStreams().values().stream()
+            .map(WindmillStreamSender::getAndResetThrottleTime)
+            .collect(toImmutableList());
+
+    return ImmutableList.<Long>builder()
+        .add(getWorkerMetadataThrottleTimer.getAndResetThrottleTime())
+        .addAll(keyedWorkStreamThrottleTimes)
+        .build();
+  }
+
+  /** Starts {@link GetWorkerMetadataStream}. */
+  @SuppressWarnings({
+    "FutureReturnValueIgnored", // ignoring Future returned from Executor.submit()
+    "nullness" // Uninitialized value of getWorkerMetadataStream is null.
+  })
+  private synchronized void startGetWorkerMetadataStream() {
+    // We only want to set and start this value once.
+    getWorkerMetadataStream.compareAndSet(
+        null,
+        streamingEngineStreamFactory.createGetWorkerMetadataStream(
+            dispatcherClient.getDispatcherStub(),
+            getWorkerMetadataThrottleTimer,
+            endpoints ->
+                consumeWorkerMetadataExecutor.submit(
+                    () -> consumeWindmillWorkerEndpoints(endpoints))));
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startBudgetRefreshThreads() {
+    scheduledBudgetRefreshExecutor.scheduleAtFixedRate(
+        () -> {
+          if (running.get() && started.get()) {
+            refreshBudget();
+          }
+        },
+        SCHEDULED_BUDGET_REFRESH_MILLIS,
+        SCHEDULED_BUDGET_REFRESH_MILLIS,
+        TimeUnit.MILLISECONDS);
+    triggeredBudgetRefreshListener.submit(this::triggeredBudgetRefresh);
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void triggeredBudgetRefresh() {
+    while (running.get() && started.get()) {
+      try {
+        // Blocks here until a value is available.
+        budgetRefreshTrigger.take();
+        refreshBudget();
+      } catch (InterruptedException e) {
+        LOG.error("Error occurred during triggered budget refresh. ", e);
+        break;
+      }
+    }
+
+    // On error or if not ready retry.
+    triggeredBudgetRefreshListener.submit(this::triggeredBudgetRefresh);
+  }
+
+  private void refreshBudget() {
+    if (isBudgetRefreshPaused.get()) {
+      return;
+    }
+
+    synchronized (this) {
+      getWorkBudgetDistributor.distributeBudget(
+          connections.get().windmillStreams().values(), totalGetWorkBudget);
+    }
+  }
+
+  private void requestBudgetRefresh() {
+    try {
+      budgetRefreshTrigger.put(true);
```

Comment:
this is going to cause this thread to block until the refresh thread takes it.

another option would be to use some sort of lock where you set value to true, and the other thread is waiting until condition is met, where then it clears value, releases lock, processes budget and then goes back to waiting.  That makes the requestBudgetRefresh entirely non-blocking which if we're going to call from grpc threads etc is likely a property we want 

com/google/common/util/concurrent/Monitor.java is perhaps one way to do that that though there may be better options.

It might be nice to structure so that refreshing is always done by a single background thread which either waits for triggers or if we need periodic evaluation can timeout and reevaluate as well.  Having a single modifier will make it easier to think through the possible races.

Reply:
updated this